### ISSUE TITLE
Can't get realm sync to work with anonymous authentication

### ISSUE THREAD

**Issue by ChaseFlorell** [realm/realm-dotnet#2490]
I'm attempting to get a list of items from a Realm db in my Xamarin Forms application. I'm pretty sure I've followed the instructions correctly, however I can't seem to see any values come in.

I'm using Realm and Realm.Fody version: `10.2.0`

I've uploaded my data with the following script

```sh
mongoimport --uri mongodb+srv://[user]:[pass]@cluster0.[wxyz].mongodb.net/[company] --collection products --type json --file products --jsonArray
```

> 1057 document(s) imported successfully. 0 document(s) failed to import.

When I go to MongoDB Atlas, I see this
[![enter image description here][1]][1]

In my App.xaml.cs constructor, I create a realm app and log in with anonymous credentials

```csharp
public const string MasterDataPartitionKey = "master_data";
ctor()
{
    var app = Realms.Sync.App.Create("my-app-id");
	app.LogInAsync(Credentials.Anonymous()).ContinueWith(task => User = task.Result);
}
```

After this, in my first ViewModel constructor (it's Rx, but isn't doing anything fancy)

```csharp
ctor()
{
	_listProductsCommand = ReactiveCommand.Create<Realm, IEnumerable<Product>>(ExecuteLoadProducts);

	_listProductsCommand.ThrownExceptions.Subscribe(x => { /* checking for errors here */ }).DisposeWith(TrashBin);
	Initialize
		.ObserveOn(RxApp.MainThreadScheduler)
                // note MasterDataPartitionKey is the value I've set on every single record's `_partitionKey` property.
		.Select(_ => new SyncConfiguration(MasterDataPartitionKey, App.User))
		.SelectMany(Realm.GetInstanceAsync)
		.Do(_ => { }, ex => {  /* checking for errors here */ })
		.ObserveOn(RxApp.MainThreadScheduler)
		.InvokeCommand(this, x => x._listProductsCommand)
		.DisposeWith(TrashBin);
}
```

And later, a simple query

```csharp
private static IEnumerable<Product> ExecuteLoadProducts(Realm realm)
{
	var data = realm.All<ProductDto>();
	var productDtos = data.ToList();
	var products = productDtos.Select(x => x.Map());
	return products;
}
```

**Expected**: 
productDtos (and products) should have a count of 1057

**Actual**: 
productDtos (and products) have a count of 0

I've looked in my Realm logs and I can see the connections and sync logs

[![enter image description here][2]][2]
[![enter image description here][3]][3]

My anonymous authentication is turned on

[![enter image description here][4]][4]

and I've made it so that all of my Collections have read access (in fact developer mode is on)

Here's an example of one of the records
[![enter image description here][5]][5]

Here's a snipped of the dto I'm trying to pull down
[![enter image description here][6]][6]

I feel as though I must be missing something simple. Can anyone see anything obvious that could be sending me sideways?

  [1]: https://i.stack.imgur.com/N07B5.png
  [2]: https://i.stack.imgur.com/asJGN.png
  [3]: https://i.stack.imgur.com/ryWRx.png
  [4]: https://i.stack.imgur.com/eDpq6.png
  [5]: https://i.stack.imgur.com/pU8vu.png
  [6]: https://i.stack.imgur.com/n3RQg.png

## Version of Realm and Tooling

- Flavor:
  - [x] MongoDB Realm
  - [ ] Local Database only
- Client SDK Version: ? Realm and Realm.Fody version: `10.2.0`
- Client OS & Version: ? iPhone 11 Pro Max / iOS 14.5

**Comment by nirinchev** [comment]
Can you modify the exception handling code to actually log the exception? It's possible some error is thrown somewhere, which is getting swallowed.

**Comment by ChaseFlorell** [comment]
I get the following logs printed

```txt
2021-07-02 09:44:46.273922-0700 cannect.iOS[74101:6783013] 2021-07-02 16:44:46.273 Info: Connection[1]: Session[1]: client_reset_config = false, Realm exists = true, async open = false, client reset = false
2021-07-02 09:44:46.357702-0700 cannect.iOS[74101:6783013] 2021-07-02 16:44:46.357 Info: Connection[1]: Connected to endpoint '54.212.159.174:443' (from '192.168.86.34:58583')
2021-07-02 09:44:47.887889-0700 cannect.iOS[74101:6782596] TransferableBytes: 3550
2021-07-02 09:44:47.888236-0700 cannect.iOS[74101:6782596] TransferredBytes: 3550
2021-07-02 09:44:47.888976-0700 cannect.iOS[74101:6782596] /Users/chaseflorell/Library/Developer/CoreSimulator/Devices/BD1DECFD-50C6-47F0-B9BD-6E2665801E66/data/Containers/Data/Application/A5997392-0B9C-4D52-9B60-85D463DA6BB2/Documents/Data/mongodb-realm/SECRET/60de39cce96e32e59cfa00d9/s_master_data.realm
2021-07-02 09:44:47.889700-0700 cannect.iOS[74101:6781735] Product Count 0
2021-07-02 09:44:51.157271-0700 cannect.iOS[74101:6782596] Active
2021-07-02 09:44:52.228605-0700 cannect.iOS[74101:6782596] 60de39cce96e32e59cfa00d9
```

With the following modification

```csharp
Initialize
	.ObserveOn(RxApp.MainThreadScheduler)
	.Select(_ => new SyncConfiguration(Data.Constants.MasterDataPartitionKey, App.User))
	.SelectMany(Realm.GetInstanceAsync)
	.ObserveOn(RxApp.MainThreadScheduler)
	.Select(LoadProductsFromRealm)
	.Subscribe(x => _productCache.AddOrUpdate(x), ex => { Console.WriteLine(ex); })
	.DisposeWith(TrashBin);
```
```csharp
private static IEnumerable<Product> LoadProductsFromRealm(Realm realm)
{
	var session = realm.GetSession();
	realm.Error += OnRealmError;
	session
		.GetProgressObservable(ProgressDirection.Download, ProgressMode.ReportIndefinitely)
		.Subscribe(onNext: p =>
		{
			Console.WriteLine($"TransferableBytes: {p.TransferableBytes}");
			Console.WriteLine($"TransferredBytes: {p.TransferredBytes}");
			Console.WriteLine(session.Path);
			Console.WriteLine(session.State);
			Console.WriteLine(session.User.Id);
		});

	var data = realm.All<ProductDto>();
	var productDtos = data.ToList();
	var products = productDtos.Select(x => x.Map());
	Console.WriteLine($"Product Count {productDtos.Count}");
	return products;
	void OnRealmError(object sender, ErrorEventArgs e)
	{
		Console.WriteLine(e.Exception);
	}
}
```

TL;DR; Nothing prints in the console for "Exceptions", only progress and count, etc

**Comment by nirinchev** [comment]
Hm, it all looks fine - can you upload a simple project that reproduces this as there's nothing obviously wrong with your code.

**Comment by ChaseFlorell** [comment]
@nirinchev  best I can do is temporarily add you to a private repo. It's a standard Xamarin Forms app, so you shouldn't have any issues running it.

note: It'll currently ONLY work on iOS.

**Comment by nirinchev** [comment]
Awesome, thanks a lot! I'll give it a shot next week and let you know what I find.

**Comment by nirinchev** [comment]
I ran your app and can confirm that the data is not there. I can see in the sync logs that only the schema is downloaded from the server and not any actual data, which likely means that for some reason the server doesn't consider those documents to be part of the requested partition. From the client-side perspective, everything appears to be in order, so can you [open a ticket](https://support.mongodb.com/welcome) with the cloud team and ask them to take a look and figure out why the serve doesn't send any documents. You can reference this issue so that they can see the preliminary analysis we've done.

**Comment by ChaseFlorell** [comment]
@nirinchev given that I am just evaluating MongoDB for our project, I have not yet signed up for a support plan. I will see if I can turn up the verbose logging on the server side and see if anything becomes apparent.
